### ISSUE TITLE
Low Efficiency caused by matrix-vector product in for loop

### ISSUE THREAD

**Issue by sgkang** [simpeg/simpeg#587]
Hi all, 

When I am running moderately big (?) problem 670 sources, and 50,000 cells (3D-IP inversion). 
It takes about 8 hrs to run 15 GN iterations, which is not too bad. 
I need to give bit of context on IP inversion. It is linear inversion, but requries sensitivity of DC problem:
**d = Gm**
For 3D, rather than generating sensitivity matrix, I store factor of system matrix, and use that to solve linear IP problem. Hence, this shoud be fast ..., but it was not than I expected. So I did few experiments, and found an issue.

Below illustrate the issue:
![screenshot from 2017-04-20 12 07 11](https://cloud.githubusercontent.com/assets/6054371/25248130/f4d28ed4-25c1-11e7-9c75-7d5f6d7118fb.png)

In the first cell above, I factorize A and solve to compute predicted data (IP), so it requires both A inverse, and back-substitutie. But in the second cell, it is already factorized, so we do not need  A inverse, but just back-substitution is required. However, it still takes 19 sec, which is not that different from 22 sec (including factorization).  So, I was curious that back subsitution takes that much time...
In the outside, I evaluated how long that it takes:

![screenshot from 2017-04-20 12 08 31](https://cloud.githubusercontent.com/assets/6054371/25248244/5ebeb12e-25c2-11e7-9552-b78ae9e11266.png)

It only took 0.52 second. But evaluation of `problem.getAderiv` did take 16 second, which is major portion of total time. So I did break a part `problem.getAderiv` as

![screenshot from 2017-04-20 12 09 16](https://cloud.githubusercontent.com/assets/6054371/25248243/5eb7dd68-25c2-11e7-9836-afa59be9847c.png)

Then figured `problem.MeSigmaDeriv` takes most of our time. 
This does require evaluation of `mesh3D.getEdgeInnerProductDeriv`, which could be broken apart:

![screenshot from 2017-04-20 12 10 18](https://cloud.githubusercontent.com/assets/6054371/25248242/5eb696f6-25c2-11e7-8aa3-d7ea705b866a.png)

Basically, the above shows matrix-vector in a for loop is the monster. 
Does anyone have good idea what is happening here, and also a good fix of this issue?
@rowanc1 @grosenkj  @jcapriot  @fourndo @lheagy @bsmithyman 

I believe most of SimPEG code can suffer from this, and solving this issue could
hugely increase efficiency of a number of SimPEG codes!

**Comment by fourndo** [comment]
Hmmm but 0.215 s * 670 src = 144 s. 
But since you are not creating the vector every time, you are doing better in 10 s already, no?
Maybe I m not getting it...

**Comment by sgkang** [comment]
oh that v in the last cell is a matrix (e.g. mesh3D.nE x 670)! So, it supposed to be ~10s. 
but, for loop

**Comment by fourndo** [comment]
Oh gotcha, yea totally. Is it possible to turn your list of vectors into a sparse matrix?

**Comment by sgkang** [comment]
Yeah, that is possible, but it could be major change in current cod bases. 
So, first I want to understand, what could makes this issue, and think about game plan to fix.

**Comment by fourndo** [comment]
I m not expert, but I would guess that calling the operator, loading it to memory and doing the product every time is the slow part. Whereas if you do it in one call, then everything goes to memory at once and done in a single operation.

**Comment by sgkang** [comment]
Thanks Dom. I expected some efficiency, but this is much greater than I thought. It is much more than my factorization, and backsubstitution time, which is ridiculous.

**Comment by sgkang** [comment]
Here's an diagnostic of the problem:

http://stackoverflow.com/questions/43546372/matrix-scipy-sparse-matrix-dense-numpy-array-multiplication-efficiency

Interesting... I just need to ask to python community. It seems scipy code needs to be changed, but 
good to recognize what is the problem, and why it is happening.

**Comment by sgkang** [comment]
Found a simple fix due to Ari's help (@ahartikainen). 
Multiplying sliced array (fotran order) to a sparse matrix, can slow down efficiency. 
And by using `.copy()`, we can get almost (?) linear efficiency b/w matrix-matrix vs matrix-vector. 

```
mesh = Mesh.TensorMesh([70, 70, 70])
A = mesh.faceDiv * mesh.faceDiv.T
rhs = np.random.randn(mesh.nC, 100)
rhs0 = rhs[:,0].copy()
%timeit out = A * rhs0
%timeit out = A * rhs[0,:]
%timeit out = A * rhs
```
Outputs are:
```
100 loops, best of 3: 4.09 ms per loop
100 loops, best of 3: 8.73 ms per loop
1 loop, best of 3: 367 ms per loop
```

Another point is posed for `mesh.aveE2CC`. When it is called there is a concatenation using
`sp.hstack`, so this makes additional computation compared to matrix-matrix vs. looping over matrix-vector. 

I would first implement this. For `Static` packages, and extend to `EM`. 
Thanks a lot Ari for your help!

**Comment by lheagy** [comment]
@sgkang: instead of doing the `.copy()` everywhere - it is likely of more benefit to the wider community if we put an issue on scipy sparse based on Jake's response in http://stackoverflow.com/questions/43546372/matrix-scipy-sparse-matrix-dense-numpy-array-multiplication-efficiency

**Comment by sgkang** [comment]
That is true @lheagy. I'll put together, and make an issue there!

**Comment by ahartikainen** [comment]
Hi, the copy thing is only for the example in StackOverflow. It has nothing to do with SimPEG being slow.

Point was that slicing multidimensional array is slower than using whole array (contiguous layout). (see code and fig below)

The problem here is probably the sparse `hstack`. It is called everytime when `mesh.aveE2CC` is called. I understand that this value can not be saved to variable, but maybe memoization could help?

In python 3 there is `lru_cache` in functools. For python 2 we probably need to create a custom memoization function.

Other options are:

- possibility to "lock" mesh so aveE2CC can be saved internally
- avoid looping (is this even possible?)
- create custom hstack, see [link here](http://stackoverflow.com/questions/6844998/is-there-an-efficient-way-of-concatenating-scipy-sparse-matrices)

The example array `A` (see code below) has size 1Mx1M. I am not sure what is the size of your 3D mesh, but if you run `hstack` 600 times it will take some time even if its smaller than the example here. It would be good idea to see what is the case with your mesh.
    
    A
    <1000000x1000000 sparse matrix of type '<class 'numpy.float64'>'
	with 2999998 stored elements in Compressed Sparse Row format>

    %timeit sp.hstack((A, A, A))
    1 loop, best of 3: 297 ms per loop

    %timeit sp.hstack((A, A, A), format='csr')
    1 loop, best of 3: 1.49 s per loop

    %timeit sp.hstack((A, A, A), format='coo')
    1 loop, best of 3: 301 ms per loop

    A_ = A.tocoo()
    A_
    <1000000x1000000 sparse matrix of type '<class 'numpy.float64'>'
	with 2999998 stored elements in COOrdinate format>
    
    %timeit sp.hstack((A_, A_, A_))
    1 loop, best of 3: 180 ms per loop

    %timeit sp.hstack((A_, A_, A_), format='csr')
    1 loop, best of 3: 1.2 s per loop

    %timeit sp.hstack((A_, A_, A_), format='coo')
    1 loops, best of 3: 181 ms per loop

taken from StackOverflow:

<img src="https://cloud.githubusercontent.com/assets/13161958/25302712/0208bfb0-274d-11e7-9f51-530adfed9457.png" width="450">

    import numpy as np
    import scipy.sparse as sp
    import matplotlib.pyplot as plt
    plt.style.use('ggplot')

    n = int(1e6)
    m = int(100)
    e = np.ones(n)
    A = sp.spdiags(np.vstack((e, e, e)), np.array([-1, 0, 1]), n, n)
    A = A.tocsr()
    u_C = np.random.randn(n,m, )
    u_F = np.asfortranarray(u_C.copy())

    col_slice_C = list(range(1,101, 5))
    times_slice_C = []
    for col_size in col_slice_C:
        t = %timeit -o -q A * u_C[:, :col_size]
        times_slice_C.append(t.best)

    col_slice_F = list(range(1,101, 5))
    times_slice_F = []
    for col_size in col_slice_F:
        t = %timeit -o -q (A * u_F[:, :col_size])
        times_slice_F.append(t.best)

    col_copy = list(range(1,101, 5))
    times_copy = []
    for col_size in col_copy:
        u_ = u_C[:, :col_size].copy()
        t = %timeit -o -q A * u_
        times_copy.append(t.best)

    fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(7,14))

    ax1.scatter(col_slice_C, np.array(times_slice_C)*1000, lw=3, label='C-slice', color='royalblue')
    ax1.scatter(col_slice_F, np.array(times_slice_F)*1000, lw=3, label='F-slice', color='orange')
    ax1.scatter(col_copy, np.array(times_copy)*1000, lw=3, label='1D copy', color='tomato')


    ax2.scatter(col_slice_C, 
                np.array(times_slice_C)*1000/np.array(col_slice_C), 
                lw=3, label='C-slice', color='royalblue')
    ax2.scatter(col_slice_F, 
                np.array(times_slice_F)*1000/np.array(col_slice_F), 
                lw=3, label='F-slice', color='orange')
    ax2.scatter(col_copy, 
                np.array(times_copy)*1000/np.array(col_copy), 
                lw=3, label='1D copy', color='tomato')

    ax2.set_ylim(0, ax2.get_ylim()[1])

    ax1.legend(fontsize=12)
    ax2.legend(fontsize=12)

    ax1.set_xlabel('slice size', fontsize=12)
    ax1.set_ylabel('time (ms)', fontsize=12)

    ax2.set_xlabel('slice size', fontsize=12)
    ax2.set_ylabel('time per col (ms)', fontsize=12)

    fig.savefig("./timing_example", dpi=200, bbox_inches='tight')

**Comment by lheagy** [comment]
Thanks @ahartikainen, to overcome repeating `sp.hstack` on every call to aveE2CC we can cache the hstacked matrix. Once the mesh is generated, the averaging matrix doesn't change. I hadn't appreciated that `sp.hstack` could be an expensive operation! I will start an issue over in discretize
